File: scripts/hibernate.hs
import Control.Concurrent (threadDelay)
import Control.Monad (liftM2)
import System.Cmd (system)
import System.Exit (exitWith)

getValue :: String -> IO Float
getValue value = do
    content <- readFile $ "/sys/class/power_supply/BAT0/energy_" ++ value
    return $ read $ unwords $ lines content

getMax :: IO Float
getMax = getValue "full"

getNow :: IO Float
getNow = getValue "now"

hibernate :: IO ()
hibernate = system "s-hibernate" >>= exitWith

toPerc :: Float -> Float -> Float
toPerc now max = now / max * 100

currentPerc :: IO Float
currentPerc = liftM2 toPerc getNow getMax

run :: Float -> IO ()
run perc | perc < 50 = hibernate
         | otherwise = print perc
                       >> threadDelay 10000000
                       >> currentPerc >>= run

main :: IO ()
main = currentPerc >>= run

File: scripts/hibernate.rs
use std::str::{FromStr};
use std::process::Command;
use std::io::Read;
use std::fs::File;
use std::path::Path;
use std::thread;

fn get_value(v: &str) -> f64 {
    let path_str = format!("/sys/class/power_supply/BAT0/energy_{}", v);
    let mut string = String::new();
    let mut file = File::open(&Path::new(&path_str))
        .ok()
        .expect(&(format!("Can't open file {}", path_str)));

    file.read_to_string(&mut string)
        .ok()
        .expect("Error reading file to the end");

    let s = string.trim();
    let n = FromStr::from_str(s);

    n.ok().expect(&format!("Error parsing str({}) -> f64", s))
}

fn get_now () -> f64 { get_value("now")  }
fn get_max () -> f64 { get_value("full") }

fn to_perc (now: f64, max: f64) -> f64 { now / max * 100.0 }

fn hibernate () {
    Command::new("hibernate")
        .spawn()
        .ok()
        .expect("Error while hibernating");
}

fn main() {
    loop {
        let curr_perc = to_perc(get_now(), get_max());
        println!("Current % = {}", curr_perc);

        if curr_perc < 50.0 {
            hibernate();
            break;
        }

        thread::sleep_ms(60_000);
    }
}
